### PR TITLE
Development test git

### PR DESCRIPTION
**Pull request by mhost39**
### Description
Test coverage for Git client 

### Related Issues
https://github.com/threefoldtech/js-ng/issues/472

### CODE REVIEW

File: tests/clients/git/test_git.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        self.info("Check if commits has been done")
+        self.assertIn(commit_msg, str(last_commit))
+
+    def test04_git_pull(self):
+        """Test case for pull from repository
+
+        **Test Scenario**
+        - Get a git client.
+        - Delet README.md file from local repository.
+        - Reset local changes.
```
[the remark this answers is not in the document]
Comment (AhmedHanafy725) [reply tests/clients/git/test_git.py:76]: As a suggesting for testing pull method
- create a file.
- commit this change.
- pull the repo.
- check that a merging commit has been done.

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]
- Please install pre-commit https://github.com/threefoldtech/js-ng/blob/development/docs/wiki/contribution.md#pre-commit

Scenarios need to be covered.
- Setting a remote url and check that the git config has been changed to the new url, and also from the property (remote_url).
- Commit changes with `add_all=False`.

File: tests/clients/git/test_git.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+from tests.base_tests import BaseTests
+
+
+class GitTests(BaseTests):
+
+    def setUp(self):
+        super().setUp()
+        self.instance_name = self.random_name()
+        
+        j.sals.process.execute("cd /tmp/ && git clone https://github.com/mhost39/test.git")
```
[the remark this answers is not in the document]
Comment (xmonader) [reply tests/clients/git/test_git.py:13]: what ensures this repo exists?

[the remark this answers is not in the document]
Comment (AhmedHanafy725) [reply tests/clients/git/test_git.py:13]: We will use jsng repo instead.

File: tests/clients/git/test_git.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+        **Test Scenario**
+        - Get a git client.
+        - Check if branch name equal main.
+        """
+        self.info("Check if branch name equal main")
+        self.assertEqual(self.git_client.branch_name, "main")
+
+    def test03_git_modifed_files(self):
+        """Test case for checking a modifed files.
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:53]:
```suggestion
        """Test case for getting the modifed files.
```

File: tests/clients/git/test_git.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        """Test case for checking git config file.
+
+        **Test Scenario**
+        - Get a git client.
+        - Read the git.config file.
+        - Check that remote_url equal to repo_url.
+        - Check that url equal to remote_url.
+        """
+        self.info("Read the git.config file")
+        git_config = j.sals.fs.read_file(f"/tmp/{self.repo_dir}/test/.git/config")
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:34]: I think repo name should be a variable in the setUp method, and used in all tests.

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]
- Please don't forget the pre-commit installation, and setting remote url scenario.

File: tests/clients/git/test_git.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        self.repo_dir = f"/tmp/{self.random_name()}"
+        self.repo_name = "js-ng"
+        self.repo_url = "https://github.com/threefoldtech/js-ng"
+        j.sals.fs.mkdir(f"{self.repo_dir}")
+        j.sals.process.execute(f"git clone {self.repo_url}", cwd=f"{self.repo_dir}")
+        self.git_client = j.clients.git.new(name=self.instance_name, path=f"{self.repo_dir}/{self.repo_name}/")
+
+    def tearDown(self):
+        j.clients.git.delete(self.instance_name)
+        j.sals.fs.rmtree(path=f"{self.repo_dir}")
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:20]: Why using f"string" when we can use `self.repo_dir` directly?

File: tests/clients/git/test_git.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        **Test Scenario**
+        - Get a git client.
+        - Get the branch name.
+        - Check branch name.
+        """
+        self.info("Get the branch name")
+        branch_name = j.sals.process.execute("git branch", cwd=f"{self.repo_dir}/{self.repo_name}")
+
+        self.info("Check branch name")
+        self.assertIn(self.git_client.branch_name, branch_name[1])
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:55]: `git branch` return all branch and maybe the current branch is not the one returned from git client.

Comment (AhmedHanafy725) [reply tests/clients/git/test_git.py:55]: `git branch` put * before the current branch.

File: tests/clients/git/test_git.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+        self.info("Commit changes")
+        self.git_client.commit(f"add {file_name}")
+
+        self.info("modify this file")
+        j.sals.fs.write_file(f"{self.repo_dir}/{self.repo_name}/{file_name}", "test modify file")
+
+        self.info("Check if file has been modifed")
+        modided_file = self.git_client.get_modified_files()
+        self.assertNotEqual(len(modided_file), 0)
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:79]: It's better to use `assertTrue` on the list.

File: tests/clients/git/test_git.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        - Create a file in repository path.
+        - Check if a new file has been added
+        """
+        self.info("Create a file in repository path")
+        file_name = self.random_name()
+        j.sals.fs.touch(f"{self.repo_dir}/{self.repo_name}/{file_name}")
+
+        self.info("Check if a new file has been added")
+        added_file = self.git_client.get_modified_files()
+        self.assertNotEqual(len(added_file), 0)
```
Comment (AhmedHanafy725) [inline tests/clients/git/test_git.py:96]: It's better to use `assertTrue` on the list.

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

File: tests/clients/git/test_git.py (hunk 6 above)
Comment (mhost39) [reply tests/clients/git/test_git.py:55]: can use git branch --show-current ?

Comment (AhmedHanafy725) [reply tests/clients/git/test_git.py:55]: Sure

File: tests/base_tests.py [hunk 9]
```diff
@@ -15,5 +16,8 @@ def tearDown(self):
     def generate_random_text():
         return j.data.idgenerator.chars(10)
 
+    def random_name(self):
+        return j.data.idgenerator.nfromchoices(10, string.ascii_letters)
```
Comment (abom) [inline tests/base_tests.py:20]: Can be replaced by

```suggestion
        return j.data.idgenerator.chars(10)
```

Comment (mhost39) [reply tests/base_tests.py:20]: some names invalid like name start with number

image

File: tests/base_tests.py [hunk 10]
```diff
@@ -15,5 +16,8 @@ def tearDown(self):
     def generate_random_text():
         return j.data.idgenerator.chars(10)
 
+    def random_name(self):
```
Comment (AhmedHanafy725) [inline tests/base_tests.py:19]: Please update the branch from development and remove this method as it is already in `base_tests`

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]
Please check the typo CI

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**